PizzaBuilder/PizzaBuilder/Program.cs

```diff
@@ -0,0 +1,176 @@
+        {
+            WhitePizza result = this._pizza;
+            this.Reset();
+            return result;
+        }
+
+        public void BuildDough()
+        {
+            this._pizza.Add("White Dough");
+        }
+
+        public void BuildSauce()
+        {
+            this._pizza.Add("Tomato Sauce");
+        }
+
+        public void BuildTopping()
+        {
+            this._pizza.Add("Cheese");
+        }
+
+    }
+
+
+    public class WheatPizzaBuilder : IPizzaBuilder
+    {
+        private WheatPizza _pizza;
+        public WheatPizzaBuilder()
+        {
+            this.Reset();
+        }
+        public void Reset()
+        {
+            this._pizza = new WheatPizza();
+        }
+        
+        public WheatPizza GetPizza()
+        {
+            WheatPizza result = this._pizza;
+            this.Reset();
+            return result;
+        }
+
+        public void BuildDough()
+        {
+            this._pizza.Add("Wheat Dough");
+        }
+
+        public void BuildSauce()
+        {
+            this._pizza.Add("Tomato Sauce");
+        }
+
+        public void BuildTopping()
+        {
+            this._pizza.Add("Cheese");
+        }
+
+    }
+
+    public class WheatPizza
+    {
+        private List<object> parts = new List<object>();
+        public void Add(string part)
+        {
+            this.parts.Add(part);
+        }
+        public string listParts()
+        {
+            string str = "";
+            for (int i = 0; i < this.parts.Count; i++)
+            {
+                if (i < this.parts.Count - 1) str += this.parts[i] + ",";
+                else str += this.parts[i] + ".";
+            }
+            return "Pizaa componenets:" + str;
+        }
+    }
+
+    public class WhitePizza
+    {
+        private List<object> parts = new List<object>();
+        public void Add(string part)
+        {
+            this.parts.Add(part);
+        }
+        public string listParts()
+        {
+            string str = "";
+            for (int i = 0; i < this.parts.Count; i++)
+            {
+                if (i < this.parts.Count - 1) str += this.parts[i] + ",";
+                else str += this.parts[i] + ".";
+            }
+            return "Pizaa componenets:" + str;
+        }
+    }
+
+    public class Director
+    {
+        private IPizzaBuilder _builder;
+        public IPizzaBuilder Builder
+        {
+            set { this._builder = value; }
+        }
+        public void WithSuase()
+        {
+            this._builder.BuildDough();
+            this._builder.BuildSauce();
+            this._builder.BuildTopping();
+        }
+        public void WithoutSuase()
+        {
+            this._builder.BuildDough();
+            this._builder.BuildSauce();
+        }
+    }
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var director = new Director();
+            var wheatPizzaBuilder = new WheatPizzaBuilder();
```

Comment:
It would be better if you used polymorphism here, without declearing two builders.

Reply:
How will the director then know what type of builder I want ?

Reply:
If you want , you can apply casting on the object to the desired builder.. you can do it this way: 
```C#
 var director = new Director();
            //var wheatPizzaBuilder = new WheatPizzaBuilder();
            IPizzaBuilder pizzaBuilder = new WheatPizzaBuilder();
            director.Builder = (WheatPizzaBuilder)pizzaBuilder;

            //Wheat Pizza With Suase
            director.WithSuase();
            Console.WriteLine(((WheatPizzaBuilder)pizzaBuilder ).GetPizza().listParts());
            //Wheat Pizza With Suase
            director.WithoutSuase();
            Console.WriteLine(((WheatPizzaBuilder) pizzaBuilder ).GetPizza().listParts());
            //Bonus 1 
            pizzaBuilder.BuildDough();
            pizzaBuilder.BuildTopping();
            Console.WriteLine(((WheatPizzaBuilder) pizzaBuilder ).GetPizza().listParts());


            // var whitePizzaBuilder = new WhitePizzaBuilder();
            director.Builder = (pizzaBuilder = new WhitePizzaBuilder());

            director.Builder = pizzaBuilder;
            //Wheat Pizza With Suase
            director.WithSuase();
            Console.WriteLine(((WhitePizzaBuilder)pizzaBuilder ).GetPizza().listParts());
            //Wheat Pizza With Suase
            director.WithoutSuase();
            Console.WriteLine(((WhitePizzaBuilder) pizzaBuilder ).GetPizza().listParts());
            //Bonus 1 
            pizzaBuilder.BuildDough();
            pizzaBuilder.BuildTopping();
            Console.WriteLine(((WhitePizzaBuilder) pizzaBuilder).GetPizza().listParts());
```
or this way:
```C#
var director = new Director();
            //var wheatPizzaBuilder = new WheatPizzaBuilder();
            IPizzaBuilder pizzaBuilder = new WheatPizzaBuilder();
            director.Builder = pizzaBuilder;

            //Wheat Pizza With Suase
            director.WithSuase();
            Console.WriteLine((pizzaBuilder as WheatPizzaBuilder).GetPizza().listParts());
            //Wheat Pizza With Suase
            director.WithoutSuase();
            Console.WriteLine((pizzaBuilder as WheatPizzaBuilder).GetPizza().listParts());
            //Bonus 1 
            pizzaBuilder.BuildDough();
            pizzaBuilder.BuildTopping();
            Console.WriteLine((pizzaBuilder as WheatPizzaBuilder).GetPizza().listParts());


            // var whitePizzaBuilder = new WhitePizzaBuilder();
            director.Builder = (pizzaBuilder = new WhitePizzaBuilder());

            director.Builder = pizzaBuilder;
            //Wheat Pizza With Suase
            director.WithSuase();
            Console.WriteLine((pizzaBuilder as WhitePizzaBuilder).GetPizza().listParts());
            //Wheat Pizza With Suase
            director.WithoutSuase();
            Console.WriteLine((pizzaBuilder as WhitePizzaBuilder).GetPizza().listParts());
            //Bonus 1 
            pizzaBuilder.BuildDough();
            pizzaBuilder.BuildTopping();
            Console.WriteLine((pizzaBuilder as WhitePizzaBuilder).GetPizza().listParts());
```
I hope I answered your question! Thank you.